modules/boulder/src/main/ts/ephox/boulder/api/FieldPresence.ts

```diff
@@ -1,65 +1,35 @@
-
-export type StrictField<T> = () => T;
-export type DefaultedThunkField<T> = (fallbackThunk: (...rest: any[]) => any) => T;
-export type AsOptionField<T> = () => T;
-export type AsDefaultedOptionThunkField<T> = (fallbackThunk: (...rest: any[]) => any) => T;
-export type MergeWithThunkField<T> = (baseThunk: (...rest: any[]) => any) => T;
-export interface FieldPresenceAdt {
-  fold: <T>(
-    strict: StrictField<T>,
-    defaultedThunk: DefaultedThunkField<T>,
-    asOption: AsOptionField<T>,
-    asDefaultedOptionThunk: AsDefaultedOptionThunkField<T>,
-    mergeWithThunk: MergeWithThunkField<T>
-  ) => T;
-  match: <T>(branches: {
-    strict: StrictField<T>;
-    defaultedThunk: DefaultedThunkField<T>;
-    asOption: AsOptionField<T>;
-    asDefaultedOptionThunk: AsDefaultedOptionThunkField<T>;
-    mergeWithThunk: MergeWithThunkField<T>;
-  }) => T;
-  log: (label: string) => void;
+import { Fun } from '@ephox/katamari';
+
+export const enum FieldType {
+  Strict = 'strict',
+  DefaultedThunk = 'defaultedThunk',
+  Option = 'option',
+  DefaultedOptionThunk = 'defaultedOptionThunk',
+  MergeWithThunk = 'mergeWithThunk'
 }
 
-const adt: {
-  strict: StrictField<FieldPresenceAdt>;
-  defaultedThunk: DefaultedThunkField<FieldPresenceAdt>;
-  asOption: AsOptionField<FieldPresenceAdt>;
-  asDefaultedOptionThunk: MergeWithThunkField<FieldPresenceAdt>;
-  mergeWithThunk: MergeWithThunkField<FieldPresenceAdt>;
-} = Adt.generate([
-  { strict: [ ] },
-  { defaultedThunk: [ 'fallbackThunk' ] },
-  { asOption: [ ] },
-  { asDefaultedOptionThunk: [ 'fallbackThunk' ] },
-  { mergeWithThunk: [ 'baseThunk' ] }
-]);
-
-const defaulted = <T>(fallback: T): FieldPresenceAdt => {
-  return adt.defaultedThunk(
-    Fun.constant(fallback)
-  );
-};
-
-const asDefaultedOption = <T>(fallback: T): FieldPresenceAdt => {
-  return adt.asDefaultedOptionThunk(
-    Fun.constant(fallback)
-  );
-};
-
-const mergeWith = (base: {}): FieldPresenceAdt => {
-  return adt.mergeWithThunk(
-    Fun.constant(base)
-  );
-};
+type Callback = (...rest: any[]) => any;
+interface FieldPresenceData<D extends FieldType, T> {
+  readonly tag: D;
+  readonly process: T;
+}
 
-const strict = adt.strict;
-const asOption = adt.asOption;
-const defaultedThunk = adt.defaultedThunk;
-const asDefaultedOptionThunk = adt.asDefaultedOptionThunk;
-const mergeWithThunk = adt.mergeWithThunk;
+type StrictData = FieldPresenceData<FieldType.Strict, {}>;
+type DefaultedThunkData = FieldPresenceData<FieldType.DefaultedThunk, Callback>;
+type OptionData = FieldPresenceData<FieldType.Option, {}>;
+type DefaultedOptionThunkData = FieldPresenceData<FieldType.DefaultedOptionThunk, Callback>;
+type MergeWithThunkData = FieldPresenceData<FieldType.MergeWithThunk, Callback>;
+
+export type FieldPresenceTypes = StrictData | DefaultedThunkData | OptionData | DefaultedOptionThunkData | MergeWithThunkData;
+
+const strict = (): StrictData => ({ tag: FieldType.Strict, process: { }});
```

Comment:
Some of these could be a `Fun.constant` I believe? That does have pros and cons though but I believe it'd still be worth it, eg:
- Saves having to recreate the object on each call, however it does mean the object is always in memory.
- Minifies a little better, as it doesn't have the function wrapper.

Anyways, up to you if you do that or not :shrug:

Reply:
I should add we could also remove the thunk if we wanted to, however I do like that all of them are a function for consistency.